packages/application-shell/src/components/application-shell/application-shell.tsx

```diff
@@ -454,130 +503,113 @@ const shallowlyCoerceValues = memoize(uncoercedEnvironmentValues =>
-                          this.props.DEV_ONLY__loadAppbarMenuConfig
-                        }
-                        DEV_ONLY__loadNavbarMenuConfig={
-                          this.props.DEV_ONLY__loadNavbarMenuConfig
-                        }
-                      />
-                    )}
-                  />
-                </Switch>
-              );
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []); // <-- run only once, when component mounts
+
+  const coercedEnvironmentValues = React.useMemo(
+    () =>
+      shallowlyCoerceValues(props.environment) as Props<
+        AdditionalEnvironmentProperties
+      >['environment'],
+    [props.environment]
+  );
 
+  return (
+    <>
+      <Global
+        styles={css`
+          #app {
+            height: 100%;
+          }
+          .ReactModal__Body--open main {
+            /* When a modal is open, we should prevent the content to be scrollable */
+            overflow: hidden;
+          }
+        `}
+      />
+      <ApplicationShellProvider<AdditionalEnvironmentProperties>
+        environment={coercedEnvironmentValues}
+        trackingEventWhitelist={props.trackingEventWhitelist}
+        applicationMessages={props.applicationMessages}
+      >
+        {({ isAuthenticated }) => {
+          if (isAuthenticated)
             return (
-              <Route
-                render={({ location }) => (
-                  <Redirector
-                    to="login"
-                    location={location}
-                    environment={coercedEnvironmentValues}
-                    queryParams={{
-                      reason: LOGOUT_REASONS.UNAUTHORIZED,
-                      redirectTo: trimLeadingAndTrailingSlashes(
-                        joinPaths(window.location.origin, location.pathname)
-                      ),
-                    }}
-                  />
-                )}
-              />
+              <Switch>
+                {/* When the application redirects to this route,
+            we always force a hard redirect to the logout route of
+            the authentication service. */}
+                <Route
+                  path="/logout"
+                  render={({ location }) => (
+                    <Redirector
+                      to="logout"
+                      location={location}
+                      queryParams={{
+                        reason: LOGOUT_REASONS.USER,
+                        ...(coercedEnvironmentValues.servedByProxy
+                          ? {}
+                          : {
+                              // This will be used after being logged in,
+                              // to redirect to this location.
+                              redirectTo: window.location.origin,
+                            }),
+                      }}
+                    />
+                  )}
+                />
+                <Route
+                  render={() => (
```

Comment:
Nit: I think we can just use children as a component here to avoid a the render prop.